Kind connect (#107)
microsoft/azure_arc: azure_arc_k8s_jumpstart/docs/local_kind.md

# Overview

The following README will guide you on how to use [kind](https://kind.sigs.k8s.io/) to run a Kubernetes cluster locally and connect it as an Azure Arc enabled Kubernetes cluster resource.

# Prerequisites

* Clone this repo

    ```terminal
    git clone https://github.com/microsoft/azure_arc.git
    ```
    
* [Install or update Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). **Azure CLI should be running version 2.7** or later. Use ```az --version``` to check your current installed version.

* [Install and Set Up kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)

* [Install Helm 3](https://helm.sh/docs/intro/install/)

* Kind leverages Docker to run the Kubernetes nodes. You will need to install Docker locally:
  * If you are a Windows user, install [Docker Desktop](https://www.docker.com/products/docker-desktop). You can also use the [Chocolatey package](https://chocolatey.org/packages/docker-desktop) to automate the installation.

>   * If you are a Mac User, install [Docker Desktop for Mac](https://docs.docker.com/docker-for-mac/).

Reviewer: ```suggestion
  * If you are a MacOS User, install [Docker Desktop for Mac](https://docs.docker.com/docker-for-mac/).
```

* Enable subscription for two providers for Azure Arc enabled Kubernetes<br> 
  Registration is an asynchronous process, and registration may take approximately 10 minutes.
  ```bash
  az provider register --namespace Microsoft.Kubernetes
  Registering is still on-going. You can monitor using 'az provider show -n Microsoft.Kubernetes'

  az provider register --namespace Microsoft.KubernetesConfiguration
  Registering is still on-going. You can monitor using 'az provider show -n Microsoft.KubernetesConfiguration'
  ```
  You can monitor the registration process with the following commands:
  ```bash
  az provider show -n Microsoft.Kubernetes -o table
 
  az provider show -n Microsoft.KubernetesConfiguration -o table
  ```

* Install the Azure Arc for Kubernetes CLI extensions ***connectedk8s*** and ***k8sconfiguration***:

  ```bash
  az extension add --name connectedk8s
  az extension add --name k8sconfiguration
  ```

 
# Deployment

* Install kind
  * On Linux: 

    ```bash
    curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.9.0/kind-linux-amd64
    chmod +x ./kind
    sudo mv ./kind /usr/local/bin/kind
    ```

>   * On Mac:

Reviewer: ```suggestion
  * On MacOS:
```

    ```bash
    brew install kind
    ```
  * On Windows:
       ```powershell
       choco install kind
       ```

> 

Reviewer: ```suggestion

```

* Navigate to the folder that has the kind cluster definition.
  ```bash
  cd azure_arc/azure_arc_k8s_jumpstart/kind
  ```

> 

Reviewer: Add a section talking about the config yaml file. This will help provide some context.

Author: will add.

* Create the kind cluster
    ```bash
    kind create cluster --config kind_cluster.yaml
    ```
    ![](../img/kind_connect/01.png)

> 

Reviewer: add --name and --kubeconfig flags. 
Update screenshot accordingly

Author: will add

Author: Would you be OK with NOT adding --kube-config? The default should work for most people, and given we're making this universal across Windows/Linux/MacOS users, specifying the --kube-config flag would require another double command.

Reviewer: Yes good point. Let's till have this as an explanation in a bullet point

Author: OK. Will update this.

* Verify your cluster was created successfully and you can access the cluster using `kubectl`.
    ```bash
    kubectl get nodes
    ```
    ![](../img/kind_connect/02.png)

# Connecting to Azure Arc

Now that you have a running kind cluster, lets connect the kind cluster to Azure Arc.
  * Login to previously created [***Service Principal***](#prerequisites) <br>
    ```bash
    az login --service-principal -u mySpnClientId -p mySpnClientSecret --tenant myTenantID
    ```

 * Create a resource group<br> 
   ```bash

>    az group create --name arckinddemo -l EastUS -o table

Reviewer: ```suggestion
   az group create --name Arc-kind-Demo -l EastUS -o table
```
This is to align with the rest of the repo. I am aware that there are some exceptions but this is the general naming convention.  
Also, update the screenshot accordingly.

   ```

>    **Note: Azure Arc enabled Kubernetes is currently supported in *East US* and *West Europe***

Reviewer: Lets make sure we are dropping this once we go Public Preview

> * Deploy the Arc binaries using Azure CLI:
>   ```bash
>   az connectedk8s connect -n arckinddemo -g arckinddemo
>   ```

Reviewer: ```suggestion
* Onboard the cluster to Azure Arc using Azure CLI:
  ```bash
  az connectedk8s connect --name Arc-Kind-Demo --resource-group Arc-kind-Demo --tags 'Project=jumpstart_azure_arc_k8s'
  ```
``` 
 
This is to align with the rest of the repo. I am aware that there are some exceptions but this is the general naming convention.
Also, update the screenshots.

Reviewer: This is to align with the rest of the repo. I am aware that there are some exceptions but this is the general naming convention.
Also, update the screenshots.

Author: Will change. Can't accept this suggestion as is as name shouldn't be capitalized. But will adapt in updated PR.

Reviewer: You mean Arc cluster name cannot be capitalized? 
https://github.com/microsoft/azure_arc/raw/master/azure_arc_k8s_jumpstart/img/aks_arm_template/10.png

Author: Yes it can. But kind shouldn't be capitalized. Check the last update.

> * Deploy the Arc binaries using Azure CLI:
>   ```bash
>   az connectedk8s connect -n arckinddemo -g arckinddemo
>   ```
> 

Reviewer: ```suggestion
* Onboard the cluster to Azure Arc using Azure CLI:
  ```bash
  az connectedk8s connect --name Arc-Kind-Demo --resource-group Arc-kind-Demo --tags 'Project=jumpstart_azure_arc_k8s'
  ```
```
This is to align with the rest of the repo. I am aware that there are some exceptions but this is the general naming convention.
Also, update the screenshots.

* Upon completion, you will have your kind cluster connected as a new Azure Arc Kubernetes cluster resource in a new Resource Group.

  ![](../img/kind_connect/03.png)

  ![](../img/kind_connect/04.png)

  ![](../img/kind_connect/05.png)

# Delete the deployment

* In Azure, the most straightforward way is to delete the cluster or the Resource Group via the Azure Portal or through the CLI.

    ```bash

>     az group delete --name arckinddemo

Reviewer: ```suggestion
    az group delete --name Arc-kind-Demo
```

>     ```bash
>     az group delete --name arckinddemo
>     ```

Reviewer: ```suggestion
    ```bash
    az group delete --name Arc-kind-Demo
    ```

```

* To delete the kind cluster locally, use the following command:

>     ```bash
>     kind delete cluster
>     ```

Reviewer: ```suggestion
    ```bash
    kind delete cluster
    ```

```

  ![](../img/kind_connect/06.png)

  ![](../img/kind_connect/07.png)

> 

Reviewer: ```suggestion

```

  Registering is still on-going. You can monitor using 'az provider show -n Microsoft.KubernetesConfiguration'
  ```
  You can monitor the registration process with the following commands:
  ```bash
  az provider show -n Microsoft.Kubernetes -o table
 
  az provider show -n Microsoft.KubernetesConfiguration -o table
  ```

* Install the Azure Arc for Kubernetes CLI extensions ***connectedk8s*** and ***k8sconfiguration***:

  ```bash
  az extension add --name connectedk8s
  az extension add --name k8sconfiguration
  ```

 
# Deployment

* Install kind
  * On Linux: 

    ```bash
    curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.9.0/kind-linux-amd64
    chmod +x ./kind
    sudo mv ./kind /usr/local/bin/kind
    ```
  * On MacOS:

    ```bash
    brew install kind
    ```
  * On Windows:
       ```powershell
       choco install kind
       ```

* Navigate to the folder that has the kind cluster definition.
  ```bash
  cd azure_arc/azure_arc_k8s_jumpstart/kind
  ```

* Create the kind cluster. We are using a configuration file called `kind_cluster.yaml` to specify our cluster configuration. This will create a 3 node cluster, with 1 master node and 2 worker nodes.

>   By default, kind will store the kubeconfig file used to connect to your cluster in the ~/.kube directory. If you want to use a custom directory to store the kubeconfig file, use the `--kube-config` flag.
> 

Reviewer: ```suggestion
  
**Note**: By default, kind will store the kubeconfig file used to connect to your cluster in the ~/.kube directory. If you want to use a custom directory to store the kubeconfig file, use the `--kube-config` flag. 
If you did chose a specific location for the cluster's *kubeconfig* file make sure you are exporting its location as an environment variable using the `export KUBECONFIG=kubeconfig location` or in Windows, add this location to your PATH. 
    

```
